ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1241 @@
+        return 0;
+
+    if (qc->stream0 == NULL)
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+
+    if (!quic_read_actual(qc, qc->stream0, buf, len, bytes_read, peek))
         return 0;
-    qc->ackm = ackm;
+
+    if (*bytes_read > 0) {
+        /*
+         * Even though we succeeded, tick the reactor here to ensure we are
+         * handling other aspects of the QUIC connection.
+         */
+        ossl_quic_reactor_tick(ossl_quic_channel_get_reactor(qc->ch));
+        return 1;
+    } else if (blocking_mode(qc)) {
+        /*
+         * We were not able to read anything immediately, so our stream
+         * buffer is empty. This means we need to block until we get
+         * at least one byte.
+         */
+        args.qc         = qc;
+        args.stream     = qc->stream0;
+        args.buf        = buf;
+        args.len        = len;
+        args.bytes_read = bytes_read;
+        args.peek       = peek;
+
+        res = block_until_pred(qc, quic_read_again, &args, 0);
+        if (res <= 0) {
+            if (!ossl_quic_channel_is_active(qc->ch))
+                return QUIC_RAISE_NON_NORMAL_ERROR(qc, SSL_R_PROTOCOL_IS_SHUTDOWN, NULL);
+            else
+                return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_INTERNAL_ERROR, NULL);
+        }
+    }
+
     return 1;
 }
 
-OSSL_ACKM *ossl_quic_conn_set_akcm(QUIC_CONNECTION *qc)
+int ossl_quic_read(SSL *s, void *buf, size_t len, size_t *bytes_read)
+{
+    return quic_read(s, buf, len, bytes_read, 0);
+}
+
+int ossl_quic_peek(SSL *s, void *buf, size_t len, size_t *bytes_read)
+{
+    return quic_read(s, buf, len, bytes_read, 1);
+}
+
+/*
+ * SSL_pending
+ * -----------
+ */
+size_t ossl_quic_pending(const SSL *s)
+{
+    const QUIC_CONNECTION *qc = QUIC_CONNECTION_FROM_CONST_SSL(s);
+    size_t avail = 0;
+    int fin = 0;
+
+    if (!expect_quic_conn(qc))
+        return 0;
+
+    if (qc->stream0 == NULL || qc->stream0->rstream == NULL)
+        /* Cannot raise errors here because we are const, just fail. */
+        return 0;
+
+    if (!ossl_quic_rstream_available(qc->stream0->rstream, &avail, &fin))
+        return 0;
+
+    return avail;
+}
+
+/*
+ * QUIC Front-End I/O API: SSL_CTX Management
+ * ==========================================
+ */
+
+long ossl_quic_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
+{
+    switch (cmd) {
+    default:
+        return 0;
+    }
+}
+
+long ossl_quic_callback_ctrl(SSL *s, int cmd, void (*fp) (void))
+{
+    return 0;
+}
+
+long ossl_quic_ctx_callback_ctrl(SSL_CTX *ctx, int cmd, void (*fp) (void))
+{
+    return 0;
+}
+
+QUIC_CONNECTION *ossl_quic_conn_from_ssl(SSL *ssl)
 {
-    return qc != NULL ? qc->ackm : NULL;
+    return QUIC_CONNECTION_FROM_SSL(ssl);
+}
+
+int ossl_quic_renegotiate_check(SSL *ssl, int initok)
+{
+    /* We never do renegotiation. */
+    return 0;
+}
+
+/*
+ * This is the subset of TLS1.3 ciphers which can be used with QUIC and which we
+ * actually support.
```

Comment:
Why is TLS1_3_RFC_AES_128_CCM_SHA256 unsupported?

Reply:
Because I didn't support it in the record layer yet. It wouldn't be hard, just a question of test cases mainly...

Reply:
OK, a TODO(QUIC) then?